docs/configure.md
# Configuring Graphistry

Administrators can add users, specify passwords, TLS/SSL, persist data across sessions, connect to databases, specify ontologies, and more. 

* For a list of many investigation-oriented options, see their [settings reference page](configure-investigation.md).

* See [update, backup, & migration instructions](update-backup-migrate.md) for preserving configurations and  data across installations.

* For advanced Python notebook and application configuration, see [PyGraphistry configuration](configure-pygraphistry.md).

## Add your team and provide API keys

See [user creation docs](user-creation.md)


## Top configuration places: data/config/custom.env, data/pivot-db/config/config.json

* Graphistry is primarily configured by setting values in `data/config/custom.env`
* Connector, ontology, and pivot configuration is optionally via `data/pivot-db/config/config.json`. Many relevant options are [detailed in a reference page](configure-investigation.md).

Between edits, restart one or all Graphistry services: `docker compose stop`  and `docker compose up -d`.

We typically recommend doing targeted and localized restarts via `docker compose stop service1 service2 ...` and `docker compose up -d --force-recreate --no-deps service1 service2 ...`. Contact staff for guidance.


## Further configuration: docker-compose.yml and Caddyfile

* More advanced administrators may edit `docker-compose.yml` .  Maintenance is easier if you never edit it.
* Custom TLS is via editing `Caddyfile`([Caddy docs](https://caddyserver.com/docs/automatic-https)), see below

## SSO

Contact staff for setting up internal SSO (OIDC: Okta, ...) or via social logins to GitHub/Google

If using an external proxy or load balancer, configure Graphistry's site domain setting (see above) and check above TLS settings

Recommendations for SSO when self-hosting:
* Setup email
* Mark SSO as site-wide
* Disallow non-SSO account creation
* Decide whether SSO users can automatically join organizations without an invitation

## TLS

We encourage everyone to use HTTPS over HTTP, especially through the automatic TLS option, for [securing authentication](authentication.md)

Graphistry supports both free automatic TLS within your server (Caddy/LetsEncrypt) and offloading to an external load balancer or proxy

As part of TLS configuration, decide whether to allow cross-origin embedding of private visualizations

### TLS Hardening

#### Caddy
If Caddy is used for TLS, several additional policies may be of interest, but only use ones that match your intended usage patterns:

```
grphstry.my-website.org, :80 {
  ...
  @mismatchedHost {
        not host grphstry.my-website.org localhost
  }
  handle @mismatchedHost {
        respond 400
  }
  ...
  reverse_proxy nginx:80 {
        # enable HSTS (1yr)
        header_down Strict-Transport-Security max-age=31536000;

        # disable clients from sniffing the media type
        header_down X-Content-Type-Options nosniff

        # clickjacking protection
        header_down X-Frame-Options SAMEORIGIN

        # keep referrer data off of HTTP connections
        header_down Referrer-Policy no-referrer-when-downgrade
  }
}
```

Note: Configuration line `header_down X-Frame-Options SAMEORIGIN` will prevent all cross-origin embedding. By default, public content can be cross-origin embedded while private content cannot. See also `COOKIE_SAMESITE` for enabling authorized private content in cross-origin embeddings.

#### Application servers

Also inform the Graphistry application servers to use secure cookies in `data/config/custom.env`:

```bash
COOKIE_SECURE=true
```

For visualizations to be embeddable in different origin sites, enable `COOKIE_SECURE` and then disable `COOKIE_SAMESITE`:

```bash
COOKIE_SAMESITE=None
```

... then restart: `docker-compose up -d --force-recreate --no-deps nexus`


### Setup free Automatic TLS

Caddy supports [free automatic TLS](https://caddyserver.com/docs/automatic-https) as long as your site meets the listed conditions.  

Sample `Caddyfile`:
```
https://*.website.org:443 {
  tls {
    max_certs 100
  }
  respond /caddy/health/ 200 {
    body "{\"success\": true}"
    close
  }
  reverse_proxy nginx:80
}
```

Note that Graphistry is internally configured via `nginx`.

### Setup free Semi-Automatic TLS

The above step may fail if your server does not allow outbound internet traffic. To work around, perform a [manual LetsEncrypt Certbot handshake for Caddy](configure-tls-caddy-manual-lets-encrypt-handshake.md).

### Setup custom certs

If you already have certificates, Caddy can use them:

* Place your certs in `./.caddy/my.crt` and `./.caddy/my.key`
* Modify `Caddyfile`:

```
https://your.site.ngo:443 {
  tls /root/.caddy/my.crt /root/.caddy/my.key
  respond /caddy/health/ 200 {
    body "{\"success\": true}"
    close
  }
  reverse_proxy nginx:80
}
```

Note the use of a fully qualified domain name in the first line, and that the file paths are for the `caddy` container's file system, not the host file system

### Setup offloading TLS from a load balancer or proxy

To enforce TLS using your own outside load balancer or proxy rather than the built-in Caddy server, we recommend:

* Setup TLS at your load balancer (ex: AWS ALB) or proxy
* Ensure http traffic is auto-upgraded (redirected) to https
* Forward (stripped) http traffic to Graphistry
* Optionally, in your `Caddyfile`'s [reverse_proxy](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy) section, add [trusted_proxies](https://caddyserver.com/docs/caddyfile/directives/reverse_proxy#trusted_proxies) so your load balancer's `X-Forwarded-{For,Proto,Host}` headers are trusted for propagation to Graphistry



To safelist typical internal IPs `192.168.0.0/16 172.16.0.0/12 10.0.0.0/8 127.0.0.1/8 fd00::/8 ::1`:

```
reverse_proxy nginx:80 {
	trusted_proxies private_ranges
}
```

If you know the specific range, you can provide that as well:

```
reverse_proxy nginx:80 {
	trusted_proxies 173.245.48.0/20 103.21.244.0/22
}
```


### Debugging TLS

Custom TLS setups often fail due to the certificate, OS, network, Caddy config, and file permissions. To perform isolated checks on each, try:

* Test the certificate
* Test a [standalone Caddy static file server](https://www.baty.net/2018/using-caddy-for-serving-static-content/)
* ... Including on another box, if OS/network issues are suspected
* Check the logs of `docker-compose logs -f -t caddy nginx`
* Test whether the containers are up and ports match via `docker-compose ps`, `curl`, and `curl` from within a docker container (so within the docker network namespace)

If problems persist, please reach out to your Graphistry counterparts. Additional workarounds are possible.


## Email

*Optional*

See the [email](email.md) section

## Site domain

*Optional*

In the Admin portal, go to Sites and change the `Domain name` to your domain, such as `graphs.acme-corp.com` 

This aids scenarios such as when using an outside proxy and ensuring that web users see the intended external domain instead of the internal one leaking through

## Reverse proxy

### Built-in proxying

Graphistry routes all public traffic through Docker container Caddy, so you generally modify Docker settings for service `caddy:` in `docker-compose.yml` (or `data/config/custom.env`)  or Caddy settings in `data/config/Caddyfile`. Further internal proxying may occur, but we do not advise manipulating internal tiers.

### External proxies

You may externally redirect traffic, such as through an AWS Load Balancer or CloudFlare

* See above TLS discussion
* Set the Site  


### Change public port

If `80` / `443` are already taken, such as when running multiple Graphistry instances on the same box, you may want to change to another port. For example, to expose public port `8888` for Graphistry's HTTP traffic instead of `80`, configure `docker-compose.yml` to map it as follows:

```yml
  caddy:
    ...
    expose:
      - "8888"
    ports:
      - 8888:80
```      

### Reuse Graphistry reverse proxy and JWT auth as a bastion for other services

You can configure the Caddy service to also reverse proxy additional services, including requiring their users to log in with Graphistry account credentials. 

For an example of both public and log-required proxies, see the [graph-app-kit sample](https://github.com/graphistry/graph-app-kit/blob/master/src/caddy/full.Caddyfile).


## Dashboards

Separately [configure the public and private dashboards](configure-dashboards.md)

## Connectors

Optionally, you can configure Graphistry to use database connectors. Graphistry will orchestrate cross-database query generation, pushing them down through the database API, and returning the combined results to the user. This means Graphistry can reuse your existing scaleout infrastructure and make its data more accessible to your users without requiring a second copy to be maintained. Some connectors further support use of the [Graphistry data bridge](bridge.md) for proxying requests between a Graphistry cloud server and an intermediate on-prem data bridge instead of directly connecting to on-prem API servers.

### Security Notes 

* Graphistry only needs `read only` access to the database
* Only one system-wide connector can be used per database per Graphistry virtual server at this time. Ex: You can have Splunk user 1 + Neo4j user 2 on one running Graphistry container, and Splunk user 3 + Neo4j user 2 on another running Graphistry container.
* [Templates](templates.md) and other embedding modes do not require further Graphistry configuration be beyond potentially API key generation. However, Graphistry implementors will still need access to external dashboards, APIs, etc., into which they'll be embedding Graphistry views.

### Get started

1. Edit lines of `data/config/custom.env` corresponding to your connector and restart Graphistry (see `.env` for commented out examples):

```bash
ES_HOST=...
SPLUNK_HOST=...
```

2. Restart `graphistry`, or at least the `pivot` service:

`docker-compose stop && docker-compose up -d` or `docker-compose stop nginx pivot && docker-compose up -d`

3. Test

* In `/pivot/connectors`, configured databases should appear in the live connectors section, and clicking the status check should turn them green
* Running a sample investigation with a database query should return results



### Example: Splunk

1. Create a restricted Splunk API user role from the Splunk Web UI

* `Settings` -> `Users` -> `Add new`
* Name: any, such as `graphistry_role`
* For `capabilities`: `rest_properties get`, `rtsearch`, `search`
* For `indexes`: Any that you want exposed to the investigtation team

2. Create a restricted Splunk API user from the Splunk Web UI

* `Settings` -> `Users` -> `Add new`
* Record their name/pwd
* Assign them to the role `graphistry_user` from step 1

3. Configure Graphistry's `custom.env` with the Splunk server and user information:

```bash
### Required
SPLUNK_HOST=splunk.acme.org
SPLUNK_USER=admin
SPLUNK_KEY=...

### Optional
#SPLUNK_SCHEME=https
#SPLUNK_PORT=8089
#SPLUNK_WEB_PORT=443
#SPLUNK_SUFFIX=/en-US
#SPLUNK_CACHE_TIMEOUT=14400
#SPLUNK_SEARCH_MAX_TIME=20
#SPLUNK_USE_PROXY=false
#SPLUNK_PROXY_KEY=...
#SPLUNK_SERVER_KEY=...
```

4. Restart the pivot service (see Get Started) and test the connector as per above. You can test the underlying Splunk API configuration by running `curl -u admin:changme  https://splunk.host.name.here:8089/services/search/jobs/export  -d search="search * | head 3" -d output_mode=csv` from your local Splunk server and your Graphistry server.

### Example: Neo4j

1. Configure Graphistry's `data/pivot-db/config/config.json` with the Neo4j server and user information:

```json
{
    "neo4j": {
        "user": "neo4j",
        "password": "myneo4juser",
        "bolt": "bolt://my.neo4j.com:4687"
    }
}
```

For faster start, or to be explicit about schema, or avoid start issues, you may choose to skip schema inference and optionally provide your own:

```json
{
    "neo4j": {
        "user": "neo4j",
        "password": "myneo4juser",
        "bolt": "bolt://my.neo4j.com:4687",
        "searchMaxTime": 5000,
        "schema": {
            "inferSchema": true,
            "labelProperties": {
                "MyLabel1": [ "my_field_1", "my_field_2" ],
                "MyLabel2": [ "my_field_1", "my_field_3" ],
                "MyLabel3": [ ]
            },
            "relationshipProperties": {
                "MY_RELN_1": [ "my_field_4" ],
                "MY_RELN_2": []
            },
            "defaultTimeIndex": {
                "nodeProperties": ["my_field2"]
            }
        }
    }
}
```


2. Restart the pivot service (see Get Started) and test the connector as per above. In-tool, you can test the connection status for the connector panel by clicking the 'check status' button for the Neo4j connector. If there is an error, you can further test your Neo4j database network connectivity by trying a Graphistry Jupyter notebook demo for Neo4j to connect to Neo4j from the notebook.


### Data bridge

In scenarios such as a Graphistry cloud server accessing on-prem API servers, an intermediate on-prem data bridge may be necessary. You can mix bridged and direct (default) connectors in the same Graphistry server. Learn more about the [Graphistry data bridge](bridge.md).

### Variants

* Give your Graphistry implementation user increased permissions so they can embed Graphistry into existing dashbboard and notification systems, such as for embedded visualizations and quicklinks into contextual [investigation templates](templates.md)* Run a [Graphistry data bridge](bridge.md), if available for your connector, which may help with cases such as firewalls preventing incoming connections from Graphistry to your database
* Run a bastion server between Graphistry and your database, such as a new Splunk search head
* Create fine-grained permissions by running multiple Graphistry virtual servers, with a new Splunk role per instance


## Ontology

See [custom ontology extensions](configure-ontology.md) and [settings reference page](configure-investigation.md) for full options. Topics include controlling:

* Map Column -> Type
* Map Type -> color, icon, size
* Map node/edge titles

## Pivots

Every connector comes with a base set of pivots. See [custom pivots](configure-custom-pivots.md) for teaching Graphistry new pivots based on existing connectors and pivots.

## Performance

See [performance tuning](performance-tuning.md)



CHANGELOG.md
# Changelog

All notable changes to this repo are documented in this file. The Graphistry server is tracked in the main [Graphistry major release history documentation](https://graphistry.zendesk.com/hc/en-us/articles/360033184174-Enterprise-Release-List-Downloads).

The changelog format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and all PyGraphistry-specific breaking changes are explictly noted here.

## [Development]

## [v2.41.8 - 2024-10-27]

### Added

* ReadTheDocs site: [https://graphistry-admin-docs.readthedocs.io/](https://graphistry-admin-docs.readthedocs.io/)
* Python endpoint

### Changed

* Reorganized documentation into thematic areas
* Update RAPIDS base image
* Update RAPIDS Python env name to `base` from `rapids`
* Updated most docker compose command references to `docker compose` from `docker-compose`

### Infra

* Sphinx port
* CI

### Fixed

* Telemetry images


graphistry/config.py
import errno
import os
from os.path import expanduser, exists, dirname, realpath, join
from configmanager import Config
import requests
from prompt_toolkit import prompt
from graphistry.widgets import revisionist_commit_history_html
import sys, json
from requests.auth import HTTPBasicAuth
from fabric.api import local, settings, hide
from jinja2 import Environment
import click
import string
import random

cwd = dir_path = dirname(realpath(__file__))

DEBUG = False
SHIPYARD_HOST = 'https://shipyard.graphistry.com'
if DEBUG:
    SHIPYARD_HOST = 'http://localhost:8000'


def config_location():
    if 'GRAPHISTRY_CONFIG_HOME' in os.environ:
        return '%s/graphistry/' % expanduser(os.environ['GRAPHISTRY_CONFIG_HOME'])
    else:
        return expanduser('~/.config/graphistry/')


def ensure_dir_exists(path):
    parent_dir = expanduser(dirname(path))
    try:
        os.makedirs(parent_dir)
    except OSError as exc:
        # ignore existing destination (py2 has no exist_ok arg to makedirs)
        if exc.errno != errno.EEXIST:
            raise


def create_config_files(filename, text):
    try:
        _file = open(filename, "w")
        _file.write(text)
        _file.close()
    except OSError as e:
        if e.errno != errno.EEXIST:
            raise


def id_generator(size=6, chars=string.ascii_uppercase + string.digits):
    return ''.join(random.choice(chars) for _ in range(size))


CONFIG_SCHEMA = {
    'user': {
        'name': '',
        'username': '',
    },
    'slug': '',
    'deployments': {},
    'default_deployment': {},
    'investigations': [],
    'registry_credentials': {},
    'vizapp_container': 'us.gcr.io/psychic-expanse-187412/graphistry/release/viz-app:942',
    'pivotapp_container': 'us.gcr.io/psychic-expanse-187412/graphistry/release/pivot-app:942',
    'is_airgapped': False,
    'api_canary': '',
    'api_secret': '',
    'graphistry_key': '',
    'graphistry_host': '',
    'es_host': '',
    'es_port': '9200',
    'es_protocol': 'http',
    'es_auth': '',
    'splunk_host': '',
    'splunk_web_port': '443',
    'splunk_api_port': '8089',
    'splunk_user': '',
    'splunk_password': '',
    'splunk_protocol': 'https',
    'ip_internal_accept_list': '',
    'http_user': '',
    'http_password_hash': '',
    's3_access': '',
    's3_secret': '',
}


class Graphistry(object):
    config = None
    config_json = None

    def __init__(self):
        self.config_file = os.path.join(config_location(), 'config.json')

    def get_config(self, schema=None):
        if not schema:
            schema = CONFIG_SCHEMA
        self.config = Config(schema=schema)

        return self.config

    def create_bcrypt_container(self):
        make_bcrypt = join(cwd, 'bootstrap/make-bcrypt-contianer.sh')
        local('sudo bash {mb}'.format(mb=make_bcrypt))

    def load_investigations(self, airgapped=False):
        """
        This will load the pivotdb test data.
        :return:
        """
        self.load_config(airgapped)
        pivot_db = '.pivot-db'
        invest_dir = join(pivot_db, 'investigations')
        pivots_dir = join(pivot_db, 'pivots')

        print(pivot_db)
        if not os.path.exists(pivot_db):
            local('mkdir -p ' + pivot_db)

        if not os.path.exists(invest_dir):
            local('mkdir -p ' + invest_dir)

        if not os.path.exists(pivots_dir):
            local('mkdir -p ' + pivots_dir)

        with hide('output', 'running', 'warnings'), settings(warn_only=True):
            local('sudo chmod -R 777 ' + pivot_db)

        for inv in self.config.investigations.value:
            try:
                print("writing investigation {0}".format(inv['name']))
                investigation = join(invest_dir, '{id}.json'.format(id=inv['json']['id']))
                with open(investigation, 'w') as outfile:
                    json.dump(inv['json'], outfile, ensure_ascii=False, indent=4, sort_keys=True)

                for piv in inv['pivots']:
                    print("writing pivot {0}".format(piv['name']))
                    pivot = join(pivots_dir, '{id}.json'.format(id=piv['json']['id']))
                    with open(pivot, 'w') as outfile:
                        json.dump(piv['json'], outfile, ensure_ascii=False, indent=4, sort_keys=True)
            except Exception as e:
                print(e)

        with hide('output', 'running', 'warnings'), settings(warn_only=True):
            local('sudo chmod -R 777 ' + pivot_db)

    def login(self):
        toolbar_quip = revisionist_commit_history_html()
        username = prompt('username: ', bottom_toolbar=toolbar_quip)
        password = prompt('password: ', bottom_toolbar=toolbar_quip, is_password=True)
        res = requests.get('{0}/api/v1/config/?format=json'.format(SHIPYARD_HOST),
                           auth=HTTPBasicAuth(username, password))

        if res.status_code == 401:
            print('Invalid Username/Password')
            sys.exit()
        elif res.status_code == 403:
            print('You do not have permissions to do this action. Ask your administrator to upgrade your account.')
            sys.exit()

        conf = res.json()['results'][0]
        schema = {
            'user': conf['user'],
            'default_deployment': conf['default_deployment'],
            'investigations': conf['default_deployment']['investigations'],
            'registry_credentials': conf['default_deployment']['registry_credentials'],
            'vizapp_container': conf['default_deployment']['vizapp_container'],
            'pivotapp_container': conf['default_deployment']['pivotapp_container'],
            'is_airgapped': False,
            'api_canary': '',
            'api_secret': '',
            'graphistry_key': '',
            'graphistry_host': '',
            'es_host': '',
            'es_port': '9200',
            'es_protocol': 'http',
            'es_auth': '',
            'splunk_host': '',
            'splunk_web_port': '443',
            'splunk_api_port': '8089',
            'splunk_protocol': 'https',    
            'splunk_user': '',
            'splunk_password': '',
            'ip_internal_accept_list': '',
            'http_user': '',
            'http_password_hash': '',
            's3_access': '',
            's3_secret': '',


        }
        self.get_config(schema)

    def load_config(self, airgapped=False):
        print("Loading Config")
        if exists(self.config_file):
            self.config = Config(
                schema=CONFIG_SCHEMA,
                load_sources=[self.config_file],
                auto_load=True,
            )
            self.config.json.load(self.config_file)
        elif airgapped:
            self.get_config(CONFIG_SCHEMA)
        else:
            self.login()
        return self.config

    def template_config(self, airgapped=False):
        toolbar_quip = revisionist_commit_history_html()
        self.load_config(airgapped)

        self.config.is_airgapped.value = airgapped

        
        # Graphistry API Key Generation
        click.secho("[graphistry] Configure API key generation. [Hash algorithm is 'aes-256-cbc'.]", fg="yellow")
        
        api_canary = prompt('Hash Canary string (enter to autogenerate): ', bottom_toolbar=toolbar_quip)
        api_secret = prompt('Your Secret string (enter to autogenerate): ', bottom_toolbar=toolbar_quip)

        if api_canary == '':
            self.config.api_canary.value = id_generator(10)

        if api_secret == '':
            self.config.api_secret.value = id_generator(10)

        #self.config.graphistry_key.value = prompt('You supplied Graphisty key: ',
        #                                           bottom_toolbar=toolbar_quip)


        # Elasticsearch
        click.secho("[graphistry] Configure connectors", fg="yellow")
        self.config.es_host.value = prompt('Your Elasticsearch Host, e.g., elk.company.com (enter to skip): ',
                                           bottom_toolbar=toolbar_quip)
        if self.config.es_host.value != '':
            self.config.es_port.value = prompt('Your Elasticsearch Port [default: 9200]: ',
                                               bottom_toolbar=toolbar_quip)
            if self.config.es_port.value == '':
                self.config.es_port.value = CONFIG_SCHEMA['es_port']

            self.config.es_protocol.value = prompt('Your Elasticsearch Protocol [default: http]: ',
                                               bottom_toolbar=toolbar_quip)
            if self.config.es_protocol.value == '':
                self.config.es_protocol.value = CONFIG_SCHEMA['es_protocol']

            self.config.es_auth.value = prompt('Your Elasticsearch auth, e.g., user:password [default: none]: ',
                                               bottom_toolbar=toolbar_quip)
            if self.config.es_auth.value == '':
                self.config.es_auth.value = CONFIG_SCHEMA['es_auth']


        # Splunk
        self.config.splunk_host.value = prompt('Your Splunk Host, e.g., www.splunk.com (enter to skip): ',
                                               bottom_toolbar=toolbar_quip)
        if self.config.splunk_host.value != '':
            
            #### DEFAULTS
            self.config.splunk_api_port.value = prompt('Your Splunk API Port [default: 8089]: ',
                bottom_toolbar=toolbar_quip)
            if self.config.splunk_api_port.value == '':
                self.config.splunk_api_port.value = CONFIG_SCHEMA['splunk_api_port']

            self.config.splunk_web_port.value = prompt('Your Splunk Web UI Port [default: 443]: ',
                bottom_toolbar=toolbar_quip)
            if self.config.splunk_web_port.value == '':
                self.config.splunk_web_port.value = CONFIG_SCHEMA['splunk_web_port']

            self.config.splunk_protocol.value = prompt('Your Splunk Protocol, e.g., https [default: https]: ',
                bottom_toolbar=toolbar_quip)
            if self.config.splunk_protocol.value == '':
                self.config.splunk_protocol.value = CONFIG_SCHEMA['splunk_protocol']


            self.config.splunk_user.value = prompt('Your Splunk Username: ',
                                                   bottom_toolbar=toolbar_quip)
            self.config.splunk_password.value = prompt('Your Splunk Password: ',
                                                       bottom_toolbar=toolbar_quip, is_password=True)

        click.secho("[graphistry] Configure networking", fg="yellow")

        # Pivotapp->ETL/Vizapp
        self.config.graphistry_host.value = prompt('Your FQDN for this deployment, including protocol [e.g., http://graphistry.yourcompany.com]: ',
                                                   bottom_toolbar=toolbar_quip)

        # Ip Whitelist
        self.config.ip_internal_accept_list.value = prompt('Your Internal IP Accept Whitelist (beyond typical RFC 1918)'
                                                           ', ex:["127.0.0.1", "10.*"]',
                                                           bottom_toolbar=toolbar_quip)

        # Http Ingress
        self.config.http_user.value = prompt('HTTP Ingress Username: ',
                                             bottom_toolbar=toolbar_quip)
        password = prompt('HTTP Ingress Password: ',
                          bottom_toolbar=toolbar_quip,
                          is_password=True)
        with hide('output', 'running', 'warnings'), settings(warn_only=True):
            self.config.http_password_hash.value = local('docker run -it bcrypt bcrypt-cli "{0}" 10'.format(password),
                                                         capture=True)

        # AWS
        self.config.s3_access.value = prompt('AWS Access Key ID (enter to skip): ', bottom_toolbar=toolbar_quip)
        if self.config.s3_access.value != '':
            self.config.s3_secret.value = prompt('AWS Access Key Secret: ', bottom_toolbar=toolbar_quip)

        self.save_config()
        self.write_configs()

    def write_configs(self):
        jenv = Environment()
        jenv.filters['jsonify'] = json.dumps
        templates = ['pivot-config.json', 'httpd-config.json', 'viz-app-config.json']
        for tmpl in templates:
            _file = open(cwd+'/templates/'+tmpl, "r")
            create_config_files(tmpl, jenv.from_string(_file.read()).render(self.config.dump_values()))
            _file.close()

    def registry_auth(self):
        # Just snag whatever credentials are in the config and make sure the key is saved.
        # docker-py doesn't seem to want to accept the JSON string as a password so this works.
        key_filename = os.path.join(expanduser('~/.config/graphistry'), '.registrykey.json')
        registry_credentials = json.dumps(dict(self.config.default_deployment.value['registry_credentials']))
        _file = open(key_filename, "w")
        _file.write(registry_credentials)
        _file.close()

        local('docker login -u _json_key -p "$(cat {0})" https://us.gcr.io | tee'.format(key_filename))

    def save_config(self):
        print("Saving Config")
        location = config_location()

        ensure_dir_exists(location)

        if self.config:
            self.config.json.dump(self.config_file, with_defaults=True)
            print("Wrote config:", self.config_file)



